fms_extras/utils/generation.py

```diff
@@ -0,0 +1,155 @@
+            A list of tensors, each tensor being for a single prompt
+        kv_cache_manager: PagedKVCacheManager
+            the paged KVCacheManager that handles management of the kv-cache for paged attention
+        max_seq_len: int
+            the max sequence length of the model
+        max_new_tokens: int
+            max tokens to generate
+        temperature: float
+            temperature of softmax when sampling
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+        cudagraphs: bool
+            if True, model is using cudagraphs and input will be padded, otherwise no cudagraphs is used and input is
+            not padded
+
+    Returns:
+    Tuple[torch.Tensor, int, float, float]
+        the resulting output tokens, the number of new tokens generated, the time to first token in seconds and the
+        total decode time
+    """
+    start_time = time.time()
+    if decode_model is None:
+        decode_model = model
+
+    bsize = len(input_ids_list)
+
+    if cudagraphs and bsize != 1:
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1"
+        )
+
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids_list])
+    n_pads_init = [max_len - seq.size(0) for seq in input_ids_list]
+    input_ids = torch.stack(
+        [F.pad(input_ids_list[i], (n_pads_init[i], 0)) for i in range(bsize)]
+    )
+
+    result = input_ids
+    next_input = input_ids
+    kwargs: MutableMapping[str, Any] = dict()
+    kwargs["cache_data"] = None
+    kwargs["use_cache"] = True
+    sequence_ids: Optional[List[int]] = None
+    block_mapping_max = ((max_len + max_new_tokens) // 16) + 1
+    for i in range(max_new_tokens):
+        if i == 1:
+            start_time = time.time()
+
+        input_ids = next_input[:, -max_seq_len:]
+
+        # compute the mask
+        if i == 0:
+            is_pad = input_ids == 0
+            mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+            kwargs["mask"] = mask.tril(diagonal=0)
+        else:
+            kwargs["mask"] = None
+
+        # get the cache data and position ids if using cache
+        if sequence_ids is None:
+            num_tokens_per_sequence = torch.count_nonzero(input_ids.T, dim=0).tolist()
+        else:
+            num_tokens_per_sequence = [1 for _ in range(input_ids.size(0))]
+
+        cache_data = kv_cache_manager.allocate_tokens(
+            num_tokens_per_sequence, sequence_ids
+        )
+
+        sequence_ids = cache_data.sequence_ids
+
+        kwargs["cache_data"] = cache_data
+        kwargs["position_ids"] = cache_data.compute_position_ids(
+            num_tokens_per_sequence
```

Comment:
I think this was a previous PR that maybe I missed when revisiting but I'm still finding this confusing. The doc for this parameter says "number of tokens to be added to each sequence" ... that implies it would be `max_new_tokens` but obviously isn't. I think we still need to clean up the comments and/or parameter names. I guess this is non-pad tokens per sequence?

Reply:
This is unrelated to max_new_tokens, it is just creating a new position_ids object based on the current sequence length (not filled) in the cache and the new token to add per decode step.

As an example:

current_sequence_lengths = [50, 99, 16]
num_tokens_per_sequence = [5, 5, 5]

This will give you position_ids

[
     [50, 51, 52, 53, 54], 
     [99, 100, 101, 102, 103],
     [16, 17, 18, 19, 20]
]

If this is prefill, the current sequence length is essentially None, and a padded position_ids is created based on the prompt length

I can add more documentation to this function if it will help? Also, technically we could just make the assumption when you allocate_tokens that the num_tokens_per_sequence should be used for this computation, in which case we can remove the position_ids computation and just have it as a part of the dataclass

Reply:
right, I know it's not really related to 'max new tokens', I'm just saying the doc implies that t is. I think we need to make the names and comments make sense.

Reply:
updated and renamed to `model_input_lengths`